### PR TITLE
L1 tp2 pvml dev v2

### PR DESCRIPTION
**Pull request by Chriisbrown**
This PR adds a CNN emulator to the vertex finder package, it follows the same style as the FastHisto emulation so creates a vertex word class instead of a vertex class. To make sure this works with the object ntupler maker I have made various changes to the L1TrackObjectNtupleMaker i.e the additional option of saving an emulated vertex and accepting the correct class when filling the ntuples

This emulated version solves the asymmetry issue of the previous NNs due to a correct treatment of the argmax which instead of just taking the first argmax and returning this as the vertex it finds all instances of the maximum value in the histogram and returns their average index instead which is what the python version does. 

There is also a minor set of changes to the track quality BDT parameters as well as some of the parameters of fasthisto so to better match what is run in firmware.

### CODE REVIEW

File: L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc [hunk 1]
```diff
@@ -2503,15 +2517,28 @@ void L1TrackObjectNtupleMaker::analyze(const edm::Event& iEvent, const edm::Even
...
-  if (L1TkPrimaryVertexHandle.isValid()) {
-      for (vtxIter = L1TkPrimaryVertexHandle->begin(); vtxIter != L1TkPrimaryVertexHandle->end(); ++vtxIter) {
-        m_pv_L1reco->push_back(vtxIter->z0());
-        m_pv_L1reco_sum->push_back(vtxIter->pt());
+  if (SaveVertexEmulation){
+    if (L1TkPrimaryVertexWordHandle.isValid()) {
+        for (vtxWordIter = L1TkPrimaryVertexWordHandle->begin(); vtxWordIter != L1TkPrimaryVertexWordHandle->end(); ++vtxWordIter) {
+          m_pv_L1reco->push_back(vtxWordIter->z0());
+          m_pv_L1reco_sum->push_back(vtxWordIter->pt());
+          std::cout << vtxWordIter->z0() << std::endl;
```
Comment (BenjaminRS) [inline L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker.cc:2525]: Should avoid cout and use edm::LogVerbatim("XYZ") instead

File: L1Trigger/VertexFinder/interface/VertexFinder.h [hunk 2]
```diff
@@ -119,6 +119,9 @@ namespace l1tVertexFinder {
                                        const std::vector<unsigned int>& counts);
     /// Histogramming algorithm + cnn position
     void CNNPVZ0Algorithm(tensorflow::Session* cnnTrkSesh = 0, tensorflow::Session* cnnPVZ0Sesh = 0, tensorflow::Session* cnnAssSesh = 0);
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/interface/VertexFinder.h:121]: In fact I think we can now change the name of this algorithm to be a bit nicer. How about NNVtx and NNVtxEmulation?

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc [hunk 3]
```diff
@@ -52,14 +52,17 @@ VertexProducer::VertexProducer(const edm::ParameterSet& iConfig)
...
+      break;
   }
 
   // Tame debug printout.
   cout.setf(ios::fixed, ios::floatfield);
   cout.precision(4);
 
   //--- Define EDM output to be written to file (if required)
-  if (settings_.vx_algo() == Algorithm::FastHistoEmulation) {
+  if ((settings_.vx_algo() == Algorithm::FastHistoEmulation) | (settings_.vx_algo() == Algorithm::NNEmulation)){
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/plugins/VertexProducer.cc:65]: This should be ||

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc [hunk 4]
```diff
@@ -175,6 +193,10 @@ void VertexProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::Event
     case Algorithm::NN:
       vf.CNNPVZ0Algorithm(cnnTrkSesh_, cnnPVZ0Sesh_, cnnAssSesh_);
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/plugins/VertexProducer.cc:194]: NNVtx

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc [hunk 5]
```diff
@@ -175,6 +193,10 @@ void VertexProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::Event
...
       break;
+
+    case Algorithm::NNEmulation:
+      vf.CNNPVZ0Emulation(cnnTrkSesh_, cnnPVZ0Sesh_, cnnAssSesh_);
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/plugins/VertexProducer.cc:198]: NNVtxEmulation

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc [hunk 6]
```diff
@@ -183,7 +205,7 @@ void VertexProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::Event
 
 
   // //=== Store output EDM track and hardware stub collections.
-  if (settings_.vx_algo() == Algorithm::FastHistoEmulation) {
+  if ((settings_.vx_algo() == Algorithm::FastHistoEmulation) | (settings_.vx_algo() == Algorithm::NNEmulation)) {
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/plugins/VertexProducer.cc:208]: | -> ||

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py [hunk 7]
```diff
@@ -31,14 +31,14 @@
         # Do track quality cuts in emulation algorithms
         EM_DoQualityCuts = cms.bool(False),
         # Track-stubs Pt compatibility cut
-        FH_DoPtComp = cms.bool(True),
+        FH_DoPtComp = cms.bool(False),
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/python/VertexProducer_cff.py:34]: Why did this change?

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py [hunk 8]
```diff
@@ -55,21 +55,21 @@
         # DBSCAN min density tracks
         DBSCANMinDensityTracks = cms.uint32(2),
         # Minimum pt of tracks used to create vertex [GeV]
-        VxMinTrackPt = cms.double(2.0),
+        VxMinTrackPt = cms.double(1.9),
         # Maximum pt of tracks used to create vertex [GeV]
-        VxMaxTrackPt = cms.double(50.0),
+        VxMaxTrackPt = cms.double(512.0),
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/python/VertexProducer_cff.py:60]: This should be 127.0 as was done here: https://github.com/cms-l1t-offline/cmssw/pull/1026/files/67c8c47351447d510790cff9a65d8a3613c077ce#diff-ae726650f7db8a5acc41e0fa85ad5b12f40282ab1b027baee3f91b6da868f8d9

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py [hunk 9]
```diff
@@ -55,21 +55,21 @@
...
-        VxMaxTrackPtBehavior = cms.int32(0),
+        VxMaxTrackPtBehavior = cms.int32(1),
         # Maximum chi2 of tracks used to create vertex
-        VxMaxTrackChi2 = cms.double(100.),
+        VxMaxTrackChi2 = cms.double(1000.),
         # Minimum number of stubs associated to a track
         VxMinNStub = cms.uint32(4),
         # Minimum number of stubs in PS modules associated to a track
-        VxMinNStubPS = cms.uint32(3),
+        VxMinNStubPS = cms.uint32(0),
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/python/VertexProducer_cff.py:72]: Why was this changed?

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py [hunk 10]
```diff
@@ -31,14 +31,14 @@
...
-        FH_DoPtComp = cms.bool(True),
+        FH_DoPtComp = cms.bool(False),
         # chi2dof < 5 for tracks with Pt > 10
         FH_DoTightChi2 = cms.bool(False),
         # FastHisto algorithm histogram parameters (min,max,width) [cm]
         # TDR settings: [-14.95, 15.0, 0.1]
         # L1TkPrimaryVertexProducer: [-30.0, 30.0, 0.09983361065]
         # Firmware: [-14.4, 14.4, 0.4]
-        FH_HistogramParameters = cms.vdouble(-30.0, 30.0, 0.09983361065),
+        FH_HistogramParameters = cms.vdouble(-14.95, 15.0, 0.1),
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/python/VertexProducer_cff.py:41]: I can't recall the reason for this changing...

File: L1Trigger/VertexFinder/src/VertexFinder.cc [hunk 11]
```diff
@@ -1276,7 +1276,6 @@ namespace l1tVertexFinder {
 
       float dZ = abs(floor(((temp_z0 + 15.)/(binWidth))) - floor(((vertices.at(nnChosenPV->first).z0() + binWidth/2 + 15.)/(binWidth))))/128.;
 
-      auto& gttTrack = gttTracks_.at(counter);
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/src/VertexFinder.cc:1279]: Why did you get rid of this line? It is looping over the gttTracks_ collection with the counter.

File: L1Trigger/VertexFinder/src/VertexFinder.cc [hunk 12]
```diff
@@ -1315,4 +1314,163 @@ namespace l1tVertexFinder {
 
   } // end of CNNPVZ0Algorithm
 
+  void VertexFinder::CNNPVZ0Emulation(tensorflow::Session* cnnTrkSesh, tensorflow::Session* cnnPVZ0Sesh, tensorflow::Session* cnnAssSesh) {
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/src/VertexFinder.cc:1317]: Also calling this NNVtxEmulation etc

File: L1Trigger/VertexFinder/src/VertexFinder.cc [hunk 13]
```diff
@@ -1315,4 +1314,163 @@ namespace l1tVertexFinder {
...
+
+    // Run PV Network:
+    tensorflow::run(cnnPVZ0Sesh, {{"hist:0", inputPV}}, {"Identity:0"}, &outputPV);
+
+    for (int i(0);i<256;++i){
+      nnOutput[i]=outputPV[0].tensor<float, 3>()(0, i, 0);
+    }
+
+
+    //Find max then find all occurances of it in histogram and average their position -> python argmax layer
```
Comment (BenjaminRS) [inline L1Trigger/VertexFinder/src/VertexFinder.cc:1405]: Is it wise to find all instances of the maximum value in the histogram and returns their average index?
- What if they are spread in z/non-continuous? Or is that extremely unlikely/impossible?

**Review by BenjaminRS — COMMENTED** [review]
Hi Chris - sorry this took so long to get to. I have finished reviewing by eye the code and have some comments. Then I will do some tests with it this week and chat to you about them.

File: L1Trigger/VertexFinder/src/VertexFinder.cc (hunk 11 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/src/VertexFinder.cc:1279]: We already access the gttTrack at L1266 with a different counter, though this bit of track association code will be removed when moving to CMSSW 12

File: L1Trigger/VertexFinder/src/VertexFinder.cc (hunk 13 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/src/VertexFinder.cc:1405]: This is how the python is currently set up, so yes it isn't ideal to have it done in this way but I thought getting the same result as  the python was the first step and then we can think a bit harder about how to deal with these spread out or non-continuous cases, based on the performance of the python models I think these are very rare events but still worth thinking a bit about this

File: L1Trigger/VertexFinder/interface/VertexFinder.h (hunk 2 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/interface/VertexFinder.h:121]: Will change

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc (hunk 3 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/plugins/VertexProducer.cc:65]: Changed

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc (hunk 4 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/plugins/VertexProducer.cc:194]: Changed

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc (hunk 5 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/plugins/VertexProducer.cc:198]: Changed

File: L1Trigger/VertexFinder/plugins/VertexProducer.cc (hunk 6 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/plugins/VertexProducer.cc:208]: Changed

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py (hunk 7 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/python/VertexProducer_cff.py:34]: The default fasthisto doesn't contain this quality control, the one shown in the TDR has no cuts on stubs or chi2 so this was the baseline we were comparing to, there is space to tune the quality cuts on FH but there hasn't been a big effort on this

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py (hunk 10 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/python/VertexProducer_cff.py:41]: These are the TDR defaults (and FW defaults) for fasthisto

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py (hunk 8 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/python/VertexProducer_cff.py:60]: Can change, what was the reason for 127, was it the largest possible value given the current FW digitization?

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py (hunk 9 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/python/VertexProducer_cff.py:72]: There is no access to stub type information currently (or ever) in the FW so a cut on PS module number wouldn't be possible

File: L1Trigger/VertexFinder/src/VertexFinder.cc (hunk 12 above)
Comment (Chriisbrown) [reply L1Trigger/VertexFinder/src/VertexFinder.cc:1317]: Changed

File: L1Trigger/VertexFinder/python/VertexProducer_cff.py (hunk 9 above)
Comment (BenjaminRS) [reply L1Trigger/VertexFinder/python/VertexProducer_cff.py:72]: Ok -- perhaps can you add that as a comment in the code as well, so others know there is no access to this in the FW now?

File: L1Trigger/VertexFinder/src/VertexFinder.cc (hunk 11 above)
Comment (BenjaminRS) [reply L1Trigger/VertexFinder/src/VertexFinder.cc:1279]: Originally the counter needed to be reset for this code to work.... but ok lets just remove it for 12 then.

File: L1Trigger/VertexFinder/src/VertexFinder.cc (hunk 13 above)
Comment (BenjaminRS) [reply L1Trigger/VertexFinder/src/VertexFinder.cc:1405]: Ok I see. Can you add as a comment in the code that the method should be revisited? I fear that it will be forgotten 😄

File: L1Trigger/VertexFinder/src/VertexFinder.cc (hunk 11 above)
Comment (BenjaminRS) [reply L1Trigger/VertexFinder/src/VertexFinder.cc:1279]: Sorry - was just compiling and saw what you meant here: yes indeed this line is erroneous (I must have been looking at a different counter yesterday). I forgot to delete it months ago.

**Comment by BenjaminRS** [comment]
Ok closing down this older PR now
